toolchain/check/interface.cpp

```diff
@@ -40,55 +40,112 @@ auto BuildAssociatedEntity(Context& context, SemIR::InterfaceId interface_id,
+  // `BindSymbolicName` or an `ImportRef` naming one.
+  {
+    auto self_binding_const_inst_id =
+        context.constant_values().GetConstantInstId(self_binding_id);
+    auto bind_name_inst = context.insts().GetAs<SemIR::BindSymbolicName>(
+        self_binding_const_inst_id);
+    CARBON_CHECK(
+        context.entity_names().Get(bind_name_inst.entity_name_id).name_id ==
+            SemIR::NameId::SelfType,
+        "Expected a Self binding, found {0}", bind_name_inst);
   }
 
-  // Add the `Self` argument. First find the `Self` binding.
-  auto self_binding =
-      context.insts().GetAs<SemIR::BindSymbolicName>(bindings[arg_ids.size()]);
-  CARBON_CHECK(
-      context.entity_names().Get(self_binding.entity_name_id).name_id ==
-          SemIR::NameId::SelfType,
-      "Expected a Self binding, found {0}", self_binding);
+  auto self_binding = context.insts().Get(self_binding_id);
+  auto self_facet_type_id = SemIR::GetTypeInSpecific(
+      context.sem_ir(), interface_specific_id, self_binding.type_id());
   // Create a facet value to be the value of `Self` in the interface.
   // TODO: Pass this in instead of creating it here. The caller sometimes
   // already has a facet value.
   auto type_inst_id = context.types().GetInstId(self_type_id);
-  auto facet_value_const_id =
+  auto self_value_const_id =
       TryEvalInst(context, SemIR::InstId::None,
-                  SemIR::FacetValue{.type_id = self_binding.type_id,
+                  SemIR::FacetValue{.type_id = self_facet_type_id,
                                     .type_inst_id = type_inst_id,
-                                    .witness_inst_id = witness_inst_id});
-  arg_ids.push_back(context.constant_values().GetInstId(facet_value_const_id));
+                                    .witness_inst_id = self_witness_id});
+  return context.constant_values().GetInstId(self_value_const_id);
+}
+
+static auto GetSelfFacet(Context& context,
+                         SemIR::SpecificId interface_specific_id,
+                         SemIR::GenericId generic_id,
+                         SemIR::TypeId self_type_id,
+                         SemIR::InstId self_witness_id) -> SemIR::InstId {
+  return GetSelfFacet(
+      context, interface_specific_id,
+      GetSelfBinding(context, interface_specific_id, generic_id), self_type_id,
+      self_witness_id);
+}
+
+static auto GetGenericArgsWithSelfType(Context& context,
+                                       SemIR::SpecificId interface_specific_id,
+                                       SemIR::GenericId generic_id,
+                                       SemIR::TypeId self_type_id,
+                                       SemIR::InstId witness_inst_id)
+    -> llvm::SmallVector<SemIR::InstId> {
+  const auto& generic = context.generics().Get(generic_id);
+  auto bindings = context.inst_blocks().Get(generic.bindings_id);
+
+  llvm::SmallVector<SemIR::InstId> arg_ids;
+  arg_ids.reserve(bindings.size());
```

Comment:
Minor, but it looks like you're fetching the generic and bindings only for reserve, when you could also use `interface_args.size() + 1` for a reservation. If you prefer this setup (presumably `bindings.size()` should be the final size), since `GetSelfSpecificForInterfaceMemberWithSelfType` also fetches the `generic`, had you considered making `generic` the parameter instead of `generic_id`? Or, maybe even make it pass in `self_specific_args.size()` and use that, e.g. `arg_ids.reserve(std::max(interface_args.size() + 1, self_specific_args_size))` to accommodate the other call.

Note, this may also be somewhat due to how `GetSelfFacet`/`GetSelfBinding` are factored -- you essentially have the fetches needed for `GetSelfBinding` repeated here.

Reply:
Rearranged to avoid using the generic to reserve.